text/0000-feature-metadata.md

```diff
@@ -0,0 +1,275 @@
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning.
+
+## `public`
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
```

Comment:
Are these accessible on the command-line?

Reply:
Do you mean what will be accepted by `cargo add foo --features`? I was thinking not, but I will clarify this

Reply:
I mean `cargo build --features non-public`

I feel like they should be.

Reply:
Maybe it would be good to allow them if explicitly specified via CLI, but present a warning. I'll update this

Reply:
If they are your own features, do we need a warning?

Now doing something like `--features clap/debug`, I could see giving a warning

Reply:
To make sure we are on the same page, does this sound accurate?

- My crate using `othercrate = { features = ["nonpublic-feature"] ... }` in Cargo.toml would be a hard error
- My crate building with `--features othercrate/nonpublic-feature` would be a warning
- My crate building with `--features mycrate-nonpubic-feature` would be allowed, no warnings/errors

Reply:
> My crate building with --features othercrate/nonpublic-feature would be a warning

This is the only case, I'm unsure of (whether to warn or error).  While I mentioned a warning, that was building off of the other comments.  I could also see it being an error.  It could be a nice escape hatch for features like `clap/debug` so people don't accidentally build using them but people can specify them.  I'm unsure how often that happens in practice though.  If we want to go with a smaller subset in this RFC, we can start with an error and change it to a warning based on feedback.  Once its a warning, its harder to go change it to an error.

Reply:
In the latest update, this is what I have:

> In general, marking a feature `public = false` should make tooling treat the
> feature as non-public API. This is described as the following:
> 
> -   The feature is always usable within the same crate:
>     -   Enablement by other features, e.g. `foo = { enables =
>         [some-private-feature] }`, is allowed
>     -   Using the feature in integration tests is allowed
>     -   Using the feature in benchmarks is allowed
> -   The feature should not be accepted by `cargo add --features`
> -   The feature should not be reported from `cargo add`'s feature output report
> -   Once `rustdoc` is able to consume feature metadata, `rustdoc` should not
>     document these features unless `--document-private-items` is specified
> -   A future tool like `cargo info` shouldn't display information about these
>     features
> -   Explicitly specifying the feature via `--features somecrate/private-feature`
>     will allow enabling a private feature that would otherwise be forbidden
> 
> Attempting to use a private feature in any of the forbidden cases should result
> in an error. Exact details of how features work will likely be refined during
> implementation and experimentation.

I think that seems to cover what the consensus is while leaving us open to some future flexibility, but let me know if adjustments are needed.

Reply:
Moved to #3487